placement/placement.go

```diff
@@ -0,0 +1,667 @@
+	}
+
+	// Add all joining cluster to the set of active clusters
+	for name, join := range changes.Joins {
+		sp.log.Infof("Joining cluster %s:%s", db.Name, name)
+		db.Clusters[name] = join.Cluster
+		db.ShardAssignments[name] = &schema.ClusterShardAssignment{}
+	}
+
+	// Go through all decomissioned clusters, remove them from the active
+	// clusters list, and add their shards to an "unowned" pool
+	for name := range changes.Decomms {
+		sp.log.Infof("Decommissioning cluster %s:%s", db.Name, name)
+		db.Clusters[name].Status = schema.ClusterStatus_DECOMMISSIONING
+
+		shards := db.ShardAssignments[name]
+		if shards == nil {
+			continue
+		}
+
+		unowned = append(unowned, shards.Shards...)
+		shardCutoffs[name] = append(shardCutoffs[name], shards.Shards...)
+		db.ShardAssignments[name].Shards = nil
+	}
+
+	// Compute active clusters, sorting the names so that assignments are deterministic (for testing)
+	var (
+		activeClusters     = make(map[string]*schema.Cluster, len(db.Clusters))
+		activeClusterNames = make([]string, 0, len(db.Clusters))
+	)
+	for name, c := range db.Clusters {
+		if c.Status != schema.ClusterStatus_ACTIVE {
+			continue
+		}
+
+		activeClusters[name] = c
+		activeClusterNames = append(activeClusterNames, name)
+	}
+	sort.Strings(activeClusterNames)
+
+	// Rebalance shards on the active clusters.
+	// pass 1, return shards on overweight clusters to the "unowned" pool
+	desiredNumShards := sp.computeDesiredNumShards(activeClusters, int(db.NumShards))
+	for _, name := range activeClusterNames {
+		var (
+			desired  = desiredNumShards[name]
+			assigned = db.ShardAssignments[name]
+			removed  []uint32
+		)
+
+		if desired >= len(assigned.Shards) {
+			continue // Will acquire more shards in pass 2
+		}
+
+		numToRemove := len(assigned.Shards) - desired
+		sp.log.Infof("%s:%s wants %d shards, has %d shards, returning %d",
+			db.Name, name, desired, len(assigned.Shards), numToRemove)
+		removed, assigned.Shards = assigned.Shards[:numToRemove], assigned.Shards[numToRemove:]
+		unowned = append(unowned, removed...)
+		shardCutoffs[name] = append(shardCutoffs[name], removed...)
+	}
+
+	// Sort the unowned shards so that assignments are deterministic (for testing)
+	sort.Sort(shardsInOrder(unowned))
+
+	// pass 2, assign shards from the "unowned" pool to underweight clusters
+	for _, name := range activeClusterNames {
+		var (
+			desired  = desiredNumShards[name]
+			assigned = db.ShardAssignments[name]
+			added    []uint32
+		)
+
+		if desired <= len(assigned.Shards) {
+			continue // NB(mmihic): Should never be less at this point
+		}
+
+		// TODO(mmihic): Maybe validate that the numToAdd is >= the number remaining
+		numToAdd := desired - len(assigned.Shards)
+		sp.log.Infof("%s:%s wants %d shards, has %d shards, pulling %d shards from %d unowned",
+			db.Name, name, desired, len(assigned.Shards), numToAdd, len(unowned))
+		added, unowned = unowned[:numToAdd], unowned[numToAdd:]
```

Comment:
Would it be best to `if numToAdd > len(unowned) { return errNumToAddGreaterThanUnowned }`?